File: subarraydivisibility.py
from collections import defaultdict

n = int(input())
arr = list(map(int, input().split()))

prefix = 0
cnt = 0
mp = defaultdict(int)
mp[0] = 1 

for num in arr:
    prefix += num
    mod = prefix % n

    if mod < 0:
        mod += n  
        
    cnt += mp[mod]
    mp[mod] += 1

print(cnt)

File: subarraydivisibility.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

int main()
{
    int n;
    cin >> n;
    vector<int> arr(n);
    for (int i = 0; i < n; ++i)
    {
        cin >> arr[i];
    }

    long long prefix = 0;
    long long cnt = 0;
    unordered_map<int, int> mp;
    mp[0] = 1;

    for (int num : arr)
    {   
        prefix += num;
        int mod = ((prefix % n) + n) % n;

        cnt += mp[mod];
        mp[mod]++;
    }

    cout << cnt << endl;
    return 0;
}
